src/mono/mono/mini/interp/interp.c

```diff
@@ -8505,9 +8561,20 @@ interp_mark_stack (gpointer thread_data, GcScanFunc func, gpointer gc_data, gboo
 	if (!context || !context->stack_start)
 		return;
 
-	// FIXME: Scan the whole area with 1 call
-	for (gpointer *p = (gpointer*)context->stack_start; p < (gpointer*)context->stack_pointer; p++)
-		func (p, gc_data);
+	if (mono_interp_opt & INTERP_OPT_PRECISE_GC) {
+		MonoLMF **lmf_addr = (MonoLMF**)info->tls [TLS_KEY_LMF_ADDR];
+		if (lmf_addr)
+			interp_mark_no_ref_slots (context, *lmf_addr);
+	}
+
+	int slot_index = 0;
+	for (gpointer *p = (gpointer*)context->stack_start; p < (gpointer*)context->stack_pointer; p++) {
+		if (context->no_ref_slots && (context->no_ref_slots [slot_index / 8] & (1 << (slot_index % 8))))
+			;// This slot is marked as no ref, we don't scan it
+		else
+			func (p, gc_data);
+		slot_index++;
```

Comment:
Just summarizing so I understand how this all works.

The way this works is that sgen calls `interp_mark_stack` twice: once during the conservative phase and once during the precise phase (if the precise phase is enabled).   This code runs during the conservative phase.  The `no_ref_slots` bits are conservative: we're either sure there's definitely no managed pointer in the slot, or we're not sure what's in the slot.  This is because sometimes we push managed pointers with `MONO_TYPE_I`.

So if we're sure the slot isn't a pointer, we don't scan it at all.  otherwise if we're not sure - we scan it, and possibly create false pinning.

So it's not important that this PR precisely tracks every single managed pointer opcode.  But if we see a slot that can't possibly contain a managed pointer, we may mark it and potentially avoid some false pinning.

Reply:
/cc @cshung FYI

Reply:
I would add for clarity that the ambiguity of `MONO_TYPE_I` is not really that relevant to the pinning story. Let's say we have two vars, one int32 and one of type object that are allocated at the same offset, since the liveness doesn't interesect. Because we mark these ref bits for the whole scope of the method, the offset will be marked as potentially containing a ref. At the moment of suspend, in theory, we could tell exactly whether the int32 var or the object var is alive at the moment of execution, or maybe none of them. This would probably reduce most of the remaining false pinning.